tensorflow_probability/python/experimental/stats/sample_stats.py

```diff
@@ -389,3 +392,152 @@ def finalize(self, state):
       mean: An estimate of the mean.
     """
     return state.mean
+
+
+RunningCentralMomentsState = collections.namedtuple(
+    'RunningCentralMomentsState',
+    'num_samples, mean, sum_exponentiated_residuals')
+
+
+class RunningCentralMoments(object):
+  """Holds metadata for and computes running central moments.
+
+  `RunningCentralMoments` will compute an arbitrary central moment in
+  streaming fashion following the formula proposed by Philippe Pebay
+  (2008) [1]. For reference, the formula we refer to is the incremental
+  version of arbitrary moments (equation 2.9). Since the algorithm computes
+  moments as a function of lower ones, even if not requested, all lower
+  moments will be computed as well. Whether all moments are returned is
+  defined by the `return_all_moments` argument in the `finalize` method.
+  Note, while any arbitrarily high central moment is theoretically supported,
+  `RunningCentralMoments` cannot guarantee numerical stability for all
+  moments.
+
+  `RunningCentralMoments` objects do not hold state information. That
+  information, which includes intermediate calculations, are held in a
+  `RunningCentralMomentsState` as returned via `initialize` and `update`
+  method calls.
+
+  #### References
+  [1]: Philippe Pebay. Formulas for Robust, One-Pass Parallel Computation of
+        Covariances and Arbitrary-Order Statistical Moments. _Technical Report
+        SAND2008-6212_, 2008.
+        https://prod-ng.sandia.gov/techlib-noauth/access-control.cgi/2008/086212.pdf
+  """
+
+  def __init__(self, shape, moment, dtype=tf.float32):
+    """Instantiates this object.
+
+    Args:
+      shape: Python `Tuple` or `TensorShape` representing the shape of
+        incoming samples.
+      moment: Integer that represents the central moment of interest.
+      dtype: Dtype of incoming samples and the resulting statistics.
+        By default, the dtype is `tf.float32`. Any integer dtypes will be
+        cast to corresponding floats (i.e. `tf.int32` will be cast to
+        `tf.float32`), as intermediate calculations should be performing
+        floating-point division.
+    """
+    self.shape = shape
+    self.moment = moment
+    if dtype is tf.int64:
+      dtype = tf.float64
+    elif dtype.is_integer:
+      dtype = tf.float32
+    self.dtype = dtype
+
+  def initialize(self):
+    """Initializes an empty `RunningCentralMomentsState`.
+
+    Returns:
+      state: `RunningCentralMomentsState` representing a stream of no
+        inputs.
+    """
+    mean_stream = RunningMean(
+        self.shape, self.dtype
+    )
+    return RunningCentralMomentsState(
+        num_samples=0,
+        mean=mean_stream.initialize().mean,
```

Comment:
Why not just keep the mean_stream here? You seem to keep reconstructing it.

Reply:
Good point. Would seem kind of weird to keep it in the state so I'm going to make it a class attribute.

Reply:
I mean... there's nothing wrong with composition. In tfp.mcmc we have 'foreign' kernel results in the outer kernel results all the time.

Reply:
Ohhh are you referring to the `RunningMeanState` and not the actual `RunningMean` object?

If you did mean the former, I thought that nesting was kind of awkward as the `RunningMeanState` is simply the mean and the number of samples. If we kept the entire state, we'd be duplicating storing the number of samples in two places.

If you mean the latter, I just thought it was weird having an actual object in the kernel results (and not another state object or set of kernel results). I could be wrong and there definitely could be precedent somewhere else in tfp I haven't stumbled upon.

Reply:
I meant the former. You're right that it's awkward. You could store the `num_samples` inside RMS exclusively (i.e. remove it from RCMS), or replace the mean in RMS state with an empty tuple. I guess what you had here isn't the worst thing either, but it just speaks to the awkwardness in this API. In FunMC I just accept that I'm going to have duplicate state in a few places (with an option to do one of the above things if I really thought it was a performance issue).

Reply:
Gotcha. Yeah, I agree just having the state works. I'll simply store the `num_samples` inside RMS.